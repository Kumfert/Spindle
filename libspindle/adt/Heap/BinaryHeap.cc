//
// BinaryHeap.h -- a specialized binary heap implementation.
//
//  $Id: BinaryHeap.cc,v 1.2 2000/02/18 01:31:47 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1997, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
// 
// This implementation is a simple binary heap.  The perhaps
// natural recursion has been replaced with looping to relieve
// all the function call overheads.  The heap also maintains
// an index array so that individual elements can be accessed in
// constant time.
// 

#include "spindle/BinaryHeap.h"

/*
 * Warning!!! This heap uses an array from 1..n   _NOT_  0..(n-1)
 *            so that the following macros are always valid.
 *
 */

#define PARENT(A)  ((A)>>1)
#define LEFT(A)    ((A)<<1)
#define RIGHT(A)   (((A)<<1) | 0x0001)

#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

BinaryHeap::BinaryHeap(const int maxSize_, const int maxItem_) 
  : maxSize(maxSize_), maxItem(maxItem_), node( maxSize_ + 1 ), map( maxItem ), nItems(0) {
}

BinaryHeap::BinaryHeap(const int maxSize_, const int maxItem_, const int values[], const int nValues)
: maxSize(maxSize_), maxItem(maxItem_), node( maxSize_ + 1 ), map( maxItem ), nItems(nValues) {
  { for (int i=0;i<nItems;i++) {
    node[i+1].first = values[i];
    node[i+1].second = i;
    map[i] = i+1;
  }}
  { for (int i=nItems/2; i>0; i--) {
    heapify(i);
  }}
}

bool
BinaryHeap::resize(const int maxSize_, const int maxItem_) {
  maxSize = maxSize_;
  maxItem = maxItem_;
  node.resize( maxSize_ + 1 );
  map.resize( maxItem_ );
  nItems = 0;
  return true;
}

int 
BinaryHeap::heapify(const int subtree) {
  int i = subtree;
  int L = LEFT(i);
  int R = RIGHT(i);

  int largest = ( (L<=nItems) && (node[L].first>node[i].first) ) ? L : i;
  largest = ( (R<=nItems) && (node[R].first>node[largest].first) ) ? R : largest;

  while ( largest != i ) { // used to be "if", but removed recursion
    // swap i and largest 
    swap( node[i], node[largest] );
    // update map 
    map[node[i].second] = i;
    map[node[largest].second] = largest;
    //     heapify(largest); // the old recursive call
    i = largest;
    L = LEFT(i);
    R = RIGHT(i);
    largest = ( (L<=nItems) && (node[L].first>node[i].first) ) ? L : i;
    largest = ( (R<=nItems) && (node[R].first>node[largest].first) ) ? R : largest;
  }
  return 0;
}


int 
BinaryHeap::extractMax() {
  if ( nItems == 0 ) { return -1; }  // if heap is empty return -1
  int max = node[1].second;        // else max is always the first in the heap

  // Now put a new node at the top of the heap
  node[1] = node[nItems];
  map[node[1].second] = 1;
  map[max] = 0;    
  nItems--;

  // But new node violates heap property, so heapify
  heapify(1);
  return max;
}

int 
BinaryHeap::insert(const int Key, const int Item) {
  if (Item > maxItem) { return -1; }
  if (nItems == maxSize) { return -1; }
  nItems++; // remember node[] goes from 1..nItems
  int i = nItems;
  int parent = PARENT(i);
  // search from the leaf up for the place to insert...
  // ... shifting others down the heap as you go
  while ( (i>1) && (node[parent].first<Key) ) {
    node[i] = node[parent];
    map[node[i].second] = i;
    i = parent;
    parent = PARENT(i);
  }
  // ASSERT:  i is the place in the heap to insert
  node[i].first = Key;
  node[i].second = Item;
  map[Item] = i;
  return 0;
}



int 
BinaryHeap::incrementPriority( const int addToKey, const int Item ) {
  if ( addToKey < 0 ) {
    return decrementPriority( -addToKey, Item );
  }
  int i = map[Item];
  if ( (i<0) || (i>nItems) ) { return -1; }
  int Key = (node[i].first += addToKey) ;
 
  // Now start searching up the tree, shifting others
  // down as you go.
  int parent = PARENT(i);
  while ( (i>1) && (node[parent].first<Key) ) {
    node[i] = node[parent];
    map[node[i].second] = i;
    i = parent;
    parent = PARENT(i);
  }
  // ASSERT:  i is the place in the heap to insert
  node[i].first = Key;
  node[i].second = Item;
  map[Item] = i;
  return 0;
}


int 
BinaryHeap::decrementPriority( const int subtractFromKey, const int Item ) {
  if ( subtractFromKey < 0 ) {
    return incrementPriority( -subtractFromKey, Item );
  }
  int i = map[Item];
  if ( (i<0) || (i>nItems) ) { return -1; }
  node[i].first -= subtractFromKey ; 
 
  // Now start searching down the tree until this item
  // is bigger than both its children
  int L = LEFT(i);
  int R = RIGHT(i);

  int largest = ( (L<=nItems) && (node[L].first>node[i].first) ) ? L : i;
  largest = ( (R<=nItems) && (node[R].first>node[largest].first) ) ? R : largest;

  while ( largest != i ) { 
    // swap i and largest 
    swap( node[i], node[largest] );
    // update map 
    map[node[i].second] = i;
    map[node[largest].second] = largest;
    i = largest;
    L = LEFT(i);
    R = RIGHT(i);
    largest = ( (L<=nItems) && (node[L].first>node[i].first) ) ? L : i;
    largest = ( (R<=nItems) && (node[R].first>node[largest].first) ) ? R : largest;
  }
  return 0;
}

void 
BinaryHeap::print( FILE* stream ) const {
  fprintf( stream, "Heap.size = %d\n", size() );
  int maxDepth = -1;
  printSubTree( stream, 1, 1, maxDepth );
}


void 
BinaryHeap::printSubTree( FILE* stream, const int root, const int depth, int& maxDepth) const {
  int L = LEFT(root);
  int R = RIGHT(root);

  if ((L > nItems) && (maxDepth == -1)) { 
    maxDepth = depth; 
   }
  if (L <= nItems) { printSubTree(stream, L, depth+1, maxDepth); }
  fprintf( stream,"%*s<%d, %d>\n", (maxDepth-depth)*5, "", node[root].first, node[root].second );
  if (R <= nItems) { printSubTree(stream, R, depth+1, maxDepth); }
}


bool
BinaryHeap::peekMax( int &Key, int &Item ) const {
  if ( nItems == 0 ) { return false; }
  Key = node[1].first;
  Item = node[1].second;
  return true;
}
