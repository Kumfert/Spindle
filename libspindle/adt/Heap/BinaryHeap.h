//
// BinaryHeap.h -- a specialized binary heap implementation.
//
//  $Id: BinaryHeap.h,v 1.2 2000/02/18 01:31:48 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1997, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
// This implementation is a simple binary heap.  The perhaps
// natural recursion has been replaced with looping to relieve
// all the function call overheads.  The heap also maintains
// an index array so that individual elements can be accessed in
// constant time.
// 

#ifndef SPINDLE_BINARY_HEAP_H_
#define SPINDLE_BINARY_HEAP_H_

#ifndef SPINDLE_H_
#include "spindle/spindle.h"
#endif 

#ifndef SPINDLE_AUTO_ARRAY_H_
#include "spindle/SharedArray.h"
#endif

#ifdef REQUIRE_OLD_CXX_HEADER_SUFFIX
#include <stdio.h>
#include "pair.h"
#include "algobase.h"  // needed for swap
#else
#include <cstdio>
#include <algorithm>
#include <utility>
using namespace std;
#endif

SPINDLE_BEGIN_NAMESPACE

class BinaryHeap {
  typedef int Item;
  typedef int Key;
  typedef pair< Key, Item > NodeType;

  int  maxSize;
  Item maxItem;

  SharedArray<NodeType> node;
  SharedArray<int> map;

  int nItems;

  int heapify(const int subtree);
  void printSubTree( FILE* stream, const int root, const int depth, int& maxDepth) const;
public:
  // constructors
  BinaryHeap(const int maxSize_, const int maxItem_);
  BinaryHeap(const int maxSize_, const int maxItem_, const int values[], const int nvals );
  ~BinaryHeap() {}

  bool resize( const int maxSize_, const int maxItem_ );
  int extractMax();
  int insert( const int Key, const int Item );
  int incrementPriority( const int addToKey, const int Item );
  int decrementPriority( const int subtractFromKey, const int Item );
  void print( FILE* stream=stdout ) const;
  const int size() const { return nItems; }
  bool peekMax( int &Key, int &Item ) const;
};

SPINDLE_END_NAMESPACE

#endif

