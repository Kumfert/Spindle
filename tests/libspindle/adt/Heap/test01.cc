//
//
//

#include "spindle/BinaryHeap.h"
#include <iostream.h>
#include <stdio.h>

#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

int main() {
  ios::sync_with_stdio();
  int maxKey = 200;
  int maxItem = 100;
  BinaryHeap heap(maxKey, maxItem);

  cout << endl << "insert <50,5>" << endl;;
  heap.insert(50, 5);
  heap.print( stdout );

  cout << endl << "insert <10,1>" << endl;;
  heap.insert(10, 1);
  heap.print( stdout );

  cout << endl << "insert <40,4>" << endl;;
  heap.insert(40, 4);
  heap.print( stdout );

  cout << endl << "insert <70,7>" << endl;;
  heap.insert(70, 7);
  heap.print( stdout );

  cout << endl << "insert <30,3>" << endl;;
  heap.insert(30, 3);
  heap.print( stdout );

  cout << endl << "insert <60,6>" << endl;;
  heap.insert(60, 6);
  heap.print( stdout );
 
  cout << endl << "insert <20,2>" << endl;;
 heap.insert(20, 2);
  heap.print( stdout );

  cout << endl << "insert <80,8>" << endl;;
  heap.insert(80, 8);
  heap.print( stdout );

  cout << endl << "increment 1 by 80" << endl;;
  heap.incrementPriority(80,1);
  heap.print( stdout );
  
  cout << endl << "increment 2 by 80" << endl;;
  heap.incrementPriority(80,2);
  heap.print( stdout );
  
  cout << endl << "increment 3 by 80" << endl;;
  heap.incrementPriority(80,3);
  heap.print( stdout );

  cout << endl << "increment 4 by 5" << endl;;
  heap.incrementPriority(5,4);
  heap.print( stdout );

  cout << endl << "increment 4 by 5" << endl;;
  heap.incrementPriority(5,4);
  heap.print( stdout );

  cout << endl << "increment 4 by 5" << endl;;
  heap.incrementPriority(5,4);
  heap.print( stdout );

  cout << endl << "increment 4 by 5" << endl;;
  heap.incrementPriority(5,4);
  heap.print( stdout );

  cout << endl << "increment 4 by 5" << endl;;
  heap.incrementPriority(5,4);
  heap.print( stdout );

  cout << endl << "remove max " << heap.extractMax() << endl;
  heap.print( stdout );

  cout << endl << "remove max " << heap.extractMax() << endl;
  heap.print( stdout );

  cout << endl << "remove max " << heap.extractMax() << endl;
  heap.print( stdout );

  cout << endl << "remove max " << heap.extractMax() << endl;
  heap.print( stdout );
  
  cout << endl << "decrement 7 by 10" << endl;
  heap.decrementPriority(10,7);
  heap.print();

  cout << endl << "decrement 4 by 5" << endl;
  heap.decrementPriority(5,4);
  heap.print();

  cout << endl << "decrement 4 by 5" << endl;
  heap.decrementPriority(5,4);
  heap.print();
}
