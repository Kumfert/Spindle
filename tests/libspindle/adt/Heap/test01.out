
insert <50,5>
Heap.size = 1
<50, 5>

insert <10,1>
Heap.size = 2
<10, 1>
     <50, 5>

insert <40,4>
Heap.size = 3
<10, 1>
     <50, 5>
<40, 4>

insert <70,7>
Heap.size = 4
<10, 1>
     <50, 5>
          <70, 7>
     <40, 4>

insert <30,3>
Heap.size = 5
<10, 1>
     <50, 5>
<30, 3>
          <70, 7>
     <40, 4>

insert <60,6>
Heap.size = 6
<10, 1>
     <50, 5>
<30, 3>
          <70, 7>
<40, 4>
     <60, 6>

insert <20,2>
Heap.size = 7
<10, 1>
     <50, 5>
<30, 3>
          <70, 7>
<40, 4>
     <60, 6>
<20, 2>

insert <80,8>
Heap.size = 8
<10, 1>
     <50, 5>
          <70, 7>
     <30, 3>
               <80, 8>
     <40, 4>
          <60, 6>
     <20, 2>

increment 1 by 80
Heap.size = 8
<50, 5>
     <70, 7>
          <80, 8>
     <30, 3>
               <90, 1>
     <40, 4>
          <60, 6>
     <20, 2>

increment 2 by 80
Heap.size = 8
<50, 5>
     <70, 7>
          <80, 8>
     <30, 3>
               <100, 2>
     <40, 4>
          <90, 1>
     <60, 6>

increment 3 by 80
Heap.size = 8
<50, 5>
     <70, 7>
          <100, 2>
     <80, 8>
               <110, 3>
     <40, 4>
          <90, 1>
     <60, 6>

increment 4 by 5
Heap.size = 8
<50, 5>
     <70, 7>
          <100, 2>
     <80, 8>
               <110, 3>
     <45, 4>
          <90, 1>
     <60, 6>

increment 4 by 5
Heap.size = 8
<50, 5>
     <70, 7>
          <100, 2>
     <80, 8>
               <110, 3>
     <50, 4>
          <90, 1>
     <60, 6>

increment 4 by 5
Heap.size = 8
<50, 5>
     <70, 7>
          <100, 2>
     <80, 8>
               <110, 3>
     <55, 4>
          <90, 1>
     <60, 6>

increment 4 by 5
Heap.size = 8
<50, 5>
     <70, 7>
          <100, 2>
     <80, 8>
               <110, 3>
     <60, 4>
          <90, 1>
     <60, 6>

increment 4 by 5
Heap.size = 8
<50, 5>
     <70, 7>
          <100, 2>
     <80, 8>
               <110, 3>
     <65, 4>
          <90, 1>
     <60, 6>

remove max 3
Heap.size = 7
<70, 7>
     <80, 8>
<50, 5>
          <100, 2>
<65, 4>
     <90, 1>
<60, 6>

remove max 2
Heap.size = 6
<70, 7>
     <80, 8>
<50, 5>
          <90, 1>
<60, 6>
     <65, 4>

remove max 1
Heap.size = 5
<60, 6>
     <70, 7>
<50, 5>
          <80, 8>
     <65, 4>

remove max 8
Heap.size = 4
<50, 5>
     <60, 6>
          <70, 7>
     <65, 4>

decrement 7 by 10
Heap.size = 4
<50, 5>
     <60, 6>
          <65, 4>
     <60, 7>

decrement 4 by 5
Heap.size = 4
<50, 5>
     <60, 6>
          <60, 4>
     <60, 7>

decrement 4 by 5
Heap.size = 4
<50, 5>
     <55, 4>
          <60, 6>
     <60, 7>
